File: 2023/src/day2/day2main.rs

use crate::day2::game::{
  Game,
  GameHand,
  BagContents,
};
use crate::day2::game_parse::parse_game_line;

pub fn day2p2(input_lines: &Vec<String>) -> u16 {
  /*
    in each game you played, what is the fewest number of cubes of each color
      that could have been in the bag to make the game possible?
  */
  let mut all_min_bags: Vec<BagContents> = Vec::new();
  let mut all_games: Vec<Game> = Vec::new();
  for line in input_lines.iter() {
    let curr_game = parse_game_line(line).unwrap();
    all_games.push(curr_game)
  }
  for curr_game in all_games.iter() {
    let mut min_bag = BagContents {
      red: 0,
      green: 0,
      blue: 0,
    };
    for curr_hand in curr_game.hands.iter() {
      if curr_hand.red > min_bag.red {
        min_bag.red = curr_hand.red;
      }
      if curr_hand.green > min_bag.green {
        min_bag.green = curr_hand.green;
      }
      if curr_hand.blue > min_bag.blue {
        min_bag.blue = curr_hand.blue;
      }
    }
    all_min_bags.push(min_bag);
  }
  let min_bag_power_set_sum = all_min_bags.iter().fold(0, |acc, curr| {
    let power_val = curr.red * curr.blue * curr.green;
    return acc + power_val;
  });

  min_bag_power_set_sum
}

pub fn day2p1(input_lines: &Vec<String>) -> u16 {
  /*
    The Elf would first like to know which games would have been possible
    if the bag contained only 12 red cubes, 13 green cubes, and 14 blue cubes?
  */
  let mut all_games: Vec<Game> = Vec::new();
  for line in input_lines.iter() {
    let curr_game = parse_game_line(line).unwrap();
    all_games.push(curr_game)
  }
  let mut possible_games: Vec<&Game> = Vec::new();
  let bag_contents = BagContents {
    red: 12,
    green: 13,
    blue: 14
  };
  for curr_game in all_games.iter() {
    if is_possible_game(&curr_game, &bag_contents) {
      possible_games.push(curr_game);
    }
  }
  let id_sum = possible_games.iter().fold(0, |acc, curr| acc + curr.game_id);
  return id_sum;

  fn is_possible_game(game: &Game, bag_contents: &BagContents) -> bool {
    return game.hands.iter().all(|hand| is_possible_hand(hand, bag_contents));
  }
  
  fn is_possible_hand(hand: &GameHand, bag_contents: &BagContents) -> bool {
    return (hand.red <= bag_contents.red)
      && (hand.green <= bag_contents.green)
      && (hand.blue <= bag_contents.blue)
    ;
  }
}

File: 2023/src/day2/day2-main.ts

import { isNumber, isString } from '../util/validate-primitives';

type GameHand = {
  red: number,
  blue: number,
  green: number,
};
type BagContents = GameHand;

type Game = {
  gameId: number,
  hands: GameHand[],
};

export function day2p2(inputLines: string[]): number {
  let allGames: Game[];
  let allMinBags: BagContents[];
  let minBagPowerSetSum: number;
  allGames = inputLines.map(parseGameLine);
  allMinBags = allGames.reduce((acc, currGame) => {
    let minBag: BagContents;
    minBag = {
      red: -Infinity,
      green: -Infinity,
      blue: -Infinity,
    };
    for(let i = 0; i < currGame.hands.length; ++i) {
      let currHand: GameHand;
      currHand = currGame.hands[i];
      minBag.red = Math.max(minBag.red, currHand.red);
      minBag.green = Math.max(minBag.green, currHand.green);
      minBag.blue = Math.max(minBag.blue, currHand.blue);
    }
    acc.push(minBag);
    return acc;
  }, [] as BagContents[]);
  minBagPowerSetSum = allMinBags.reduce((acc, currMinBag) => {
    let currPowerSetSum: number;
    currPowerSetSum = currMinBag.red * currMinBag.green * currMinBag.blue;
    return acc + currPowerSetSum;
  }, 0);
  return minBagPowerSetSum;
}

export function day2p1(inputLines: string[]): number {
  let idSum: number;
  let allGames: Game[];
  let possibleGames: Game[];
  let bagContents: BagContents;
  allGames = [];
  for(let i = 0; i < inputLines.length; ++i) {
    let inputLine: string;
    let currGame: Game;
    inputLine = inputLines[i];
    currGame = parseGameLine(inputLine);
    allGames.push(currGame);
  }
  bagContents = {
    red: 12,
    green: 13,
    blue: 14,
  };
  possibleGames = allGames.reduce((acc, currGame) => {
    if(isPossibleGame(currGame, bagContents)) {
      acc.push(currGame);
    }
    return acc;
  }, [] as Game[]);

  idSum = possibleGames.reduce((acc, currGame) => {
    return acc + currGame.gameId;
  }, 0);
  return idSum;
}

function isPossibleGame(game: Game, bagContents: BagContents): boolean {
  return game.hands.every(hand => {
    return isPossibleHand(hand, bagContents);
  });
}

function isPossibleHand(hand: GameHand, bagContents: BagContents): boolean {
  return (
    (hand.red <= bagContents.red)
    && (hand.green <= bagContents.green)
    && (hand.blue <= bagContents.blue)
  );
}

function parseGameLine(line: string): Game {
  let game: Game;
  let gameHands: GameHand[];
  let lineParts: string[];
  let gamePart: string;
  let cubesPart: string;
  let gameIdStr: string | undefined;
  let gameId: number;
  let gameHandStrs: string[];

  lineParts = line.split(':');
  gamePart = lineParts[0];
  cubesPart = lineParts[1].trim();

  gameIdStr = gamePart.split(' ').pop();
  if(
    !isString(gameIdStr)
    || !isNumber(gameId = +(gameIdStr))
  ) {
    throw new Error(`Invalid gameId: ${gameIdStr}`);
  }
  gameHandStrs = cubesPart.split(';')
    .map(handStr => {
      return handStr.trim();
    });
  gameHands = [];
  for(let i = 0; i < gameHandStrs.length; ++i) {
    let gameHandStr: string;
    let gameHand: GameHand;
    gameHandStr = gameHandStrs[i];
    gameHand = parseGameHand(gameHandStr);
    gameHands.push(gameHand);
  }
  game = {
    gameId,
    hands: gameHands,
  };
  return game;
}

function parseGameHand(gameHandStr: string): GameHand {
  let gameHand: GameHand;
  let handStrs: string[];
  handStrs = !gameHandStr.includes(',')
    ? [ gameHandStr ]
    : gameHandStr.split(',').map(cube_str => {
      return cube_str.trim();
    })
  ;
  gameHand = {
    red: 0,
    blue: 0,
    green: 0,
  };
  for(let i = 0; i < handStrs.length; ++i) {
    let handStr: string;
    let handParts: string[];
    let cubeCount: number;
    handStr = handStrs[i];
    handParts = handStr.split(' ');
    cubeCount = +handParts[0];
    switch(handParts[1]) {
      case 'red':
        gameHand.red = cubeCount;
        break;
      case 'blue':
        gameHand.blue = cubeCount;
        break;
      case 'green':
        gameHand.green = cubeCount;
        break;
      default:
        throw new Error(`Invalide hand part: ${handParts[1]}`);
    }
  }
  return gameHand;
}
